Python/compile.c

```diff
@@ -1208,14 +1208,18 @@ merge_consts_recursive(struct compiler *c, PyObject *o)
     // t is borrowed reference
     PyObject *t = PyDict_SetDefault(c->c_const_cache, key, key);
     if (t != key) {
+        // o is registered in c_const_cache.  Just use it.
         Py_INCREF(t);
         Py_DECREF(key);
         return t;
     }
 
+    // We registered o in c_const_cache.
+    // When o is a tuple or frozenset, we want to merge it's
+    // items too.
```

Comment:
Would it be possible to do it backward?

* First merge items
* Then merge the container

Reply:
I think it just make program more complicated.  We need to check the container is not registered yet before merging items.  So we need to generate key tuple containing the container object first anyway.

Reply:
Would it be possible to add the newly merged container to the cache rather than modifying the existing cache entry in-place?

Reply:
It's possible.  This commit rewrites tuple merging without in-place modify of tuple and key.
https://github.com/python/cpython/commit/2eb1471fb246a6cef9a25c6e2a985cd7f4bcaa06

But I'm not sure avoiding in-place edit is so important...
We have similar code already.
https://github.com/python/cpython/blob/master/Objects/codeobject.c#L47-L95

Reply:
https://github.com/python/cpython/blob/master/Objects/codeobject.c#L47-L95

Oh wow, this code is even worse: it modify the "immutable" arguments of PyCode_New() :-(